homeassistant/components/fyta/__init__.py

```diff
@@ -47,3 +60,39 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Migrate old entry."""
+    _LOGGER.debug("Migrating from version %s", config_entry.version)
+
+    if config_entry.version > 1:
+        # This means the user has downgraded from a future version
+        return False
+
+    if config_entry.version == 1:
+        new = {**config_entry.data}
+        if config_entry.minor_version < 2:
+            fyta = FytaConnector(
+                config_entry.data[CONF_USERNAME], config_entry.data[CONF_PASSWORD]
+            )
+            credentials: dict[str, Any] = await fyta.login()
+            await fyta.client.close()
+
+            if isinstance(credentials[CONF_EXPIRATION], datetime):
+                credentials[CONF_EXPIRATION] = credentials[CONF_EXPIRATION].isoformat()
```

Comment:
Why do we do this? Wouldn't it be better if we saved the expiration as epoch? This way we don't have to trouble ourselves with the timezones, since its all UTC

Reply:
I had to introduce this, as otherwise I had problems with the reading of the config entry on line 41:
 ```
expiration: datetime = datetime.fromisoformat(
        entry.data[CONF_EXPIRATION]
    ).astimezone(ZoneInfo(tz))
```
I assume, in case of the setup or the migration the data is in the cache and without the conversion on line 82 would be available as datetime. Consequently, line 41 throws an error. Once saved in the file, it is read as `str` and therefore I read it in with `fromisoformat`.

With regards to saving the epoch, I do not yet see how this could be realized, as one would then also have to save the starting date, otherwise we do not know when the epoch ends. Or am I missing anything?

Reply:
You know the epoch is x miliseconds after 01-01-1970?

Reply:
> You know the epoch is x miliseconds after 01-01-1970?

No, I was not aware that you are referring to this. 

Is there a difference in the result? I guess the datetime would have to be saved as `timestamp()` and then read with `fromtimestamp()`, but besides this I don't see an advantage. An advantage could be, if the whole logic would be adapted from current local time to UTC, but this would require a rewrite of the client and integration, and I'm not sure if this is reasonable in the case at hand.

Reply:
The biggest advantage is that epoch is in UTC, so you don't have the whole timezone mess as it just works

Reply:
So instead of saving `2024-12-31T10:00:00+00:00` you're saving `1735603200`.

I always find to have a love hate relationship with iso date strings. Yes, it's a little less readable, but it saves some headaches with timezones and people switching timezones probably

Reply:
I agree, that if I would start from scratch, UTC might be easier. But if you agree, I would prefer to keep it as is (and perhaps adapt the logic in a separate PR at a later stage).

Reply:
The only point I don't get is why you say it would affect the library? (If I understood correctly).

Also keep in mind, if all users now have an iso string in their entry data, a migration is required to migrate that to an epoch, so it would be benificial if we can avoid that migrationcode

Reply:
If I look at the changes, we could just replace the way to serialize the timestamps. Unless the expiration isn't always of type datetime, but then I am interested in why this type can be a different type.

Reply:
> If I look at the changes, we could just replace the way to serialize the timestamps. Unless the expiration isn't always of type datetime, but then I am interested in why this type can be a different type.

I think you are right. But in my view this goes beyond the scope of this PR (which according to my understanding should be as narrow as possible).

Reply:
To make my question a bit more explicit: Where does the timezone affect the library? I think I am missing that point here. If I understand the code we get a `datetime` object back, and we do the check if the token is still valid on the integration side, so I don't see where the lib should change.

(Oh and I am with the intentions of getting this in, I'm just trying to do my due dilligence)

Reply:
Oh wait, you are passing the tz and expiration to the lib. I am wondering why, since we always refresh in the coordinator.

Reply:
> If I understand the code we get a `datetime` object back

The datetime returned from the library is local time (`datetime.now()`). For consistency purposes this should be amended to UTC, otherwise we still have to convert.

Reply:
Right I think I see the problem. Let's do it this way.

But now the next thing I was wondering, why do we have the `isinstance(credentials[CONF_EXPIRATION], datetime)`. Can it ever be not a datetime object?

Reply:
> Oh wait, you are passing the tz and expiration to the lib. I am wondering why, since we always refresh in the coordinator.

I think you are right. This would not be needed to be passed on.

Reply:
Can we do without this check or does mypy complain then? In that case I would suggest using a `cast`

Reply:
mypy seems to be fine with the omission.